contracts/Vaults/CrabVault.sol

```diff
@@ -0,0 +1,524 @@
+    }
+
+    /// @notice Deposit funds from user into vault for the next round
+    /// @param amount Value to be deposited
+    function deposit(uint amount) external canDeposit(amount) nonReentrant whenNotPaused {
+        sUSD.safeTransferFrom(msg.sender, address(this), amount);
+
+        uint nextRound = round + 1;
+
+        // new user enters the vault
+        if (balancesPerRound[round][msg.sender] == 0 && balancesPerRound[nextRound][msg.sender] == 0) {
+            require(usersCurrentlyInVault < maxAllowedUsers, "Max amount of users reached");
+            usersPerRound[nextRound].push(msg.sender);
+            userInRound[nextRound][msg.sender] = true;
+            usersCurrentlyInVault = usersCurrentlyInVault + 1;
+        }
+
+        balancesPerRound[nextRound][msg.sender] += amount;
+
+        // update deposit state of a user
+        depositReceipts[msg.sender] = DepositReceipt(nextRound, balancesPerRound[nextRound][msg.sender]);
+
+        allocationPerRound[nextRound] += amount;
+        capPerRound[nextRound] += amount;
+
+        if (address(stakingThales) != address(0)) {
+            stakingThales.updateVolume(msg.sender, amount);
+        }
+
+        emit Deposited(msg.sender, amount);
+    }
+
+    function withdrawalRequest() external nonReentrant whenNotPaused {
+        require(vaultStarted, "Vault has not started");
+        require(!withdrawalRequested[msg.sender], "Withdrawal already requested");
+        require(balancesPerRound[round][msg.sender] > 0, "Nothing to withdraw");
+        require(balancesPerRound[round + 1][msg.sender] == 0, "Can't withdraw as you already deposited for next round");
+
+        uint nextRound = round + 1;
+        if (capPerRound[nextRound] > balancesPerRound[round][msg.sender]) {
+            capPerRound[nextRound] -= balancesPerRound[round][msg.sender];
+        }
+
+        usersCurrentlyInVault = usersCurrentlyInVault - 1;
+        withdrawalRequested[msg.sender] = true;
+        emit WithdrawalRequested(msg.sender);
+    }
+
+    /// @notice Buy in range market options from Ranged AMM
+    /// @param market address of a market
+    /// @param amount number of options to be bought
+    /// @param position to buy options for
+    function trade(
+        RangedMarket market,
+        uint amount,
+        RangedMarket.Position position
+    ) external nonReentrant whenNotPaused {
+        require(vaultStarted, "Vault has not started");
+        require(amount >= minTradeAmount, "Amount less than minimum");
+        require(position == RangedMarket.Position.In, "Invalid position");
+
+        //IPositionalMarket leftMarketContract = IPositionalMarket(market.leftMarket());
+        //IPositionalMarket rightMarketContract = IPositionalMarket(market.rightMarket());
+        (uint leftMaturity, ) = market.leftMarket().times();
+        (uint rightMaturity, ) = market.rightMarket().times();
+        require(leftMaturity < (roundStartTime[round] + roundLength), "Market time not valid");
```

Comment:
no need to check left and right market separately, we can trust that if this is a legit RangedMarket, then those two maturities are identical